### ISSUE TITLE
Can resolved paths be made root-relative or relative to the output destination?

### ISSUE THREAD

**Issue by tmanhollan** [jonaskulhanek/gulp-resolve-url#3]
Thank you for accepting my pull request (#2) to fix the problem with the options. I'm trying to accomplish something else that's not directly related to that, so I'm opening this separate issue to discuss it.

In my project, the destination path of the output is not parallel to the source paths. Without getting into the complexities of my project's layout, all of the build products are beneath a "build" directory, which causes the output to be one directory deeper than the source. But at the point in the pipeline where resolveUrl examines the paths, it has no awareness of the destination, so the resolved paths can't take the extra "build" directory into account and end up being one directory too shallow.

I've explored all of the options for sourcemap generation with gulp-sourcemaps and gulp.src and I can't see any way for the input to resolveUrl to be aware of the destination path. This is what led me to explore the absolute option, which does resolve the paths to the proper location in my filesystem, but an absolute filesystem path is not useful. If it were an absolute http path, that would work, but that too requires some additional context and is still not ideal, for a number of reasons.

I have two ideas for how resolveUrl could be modified to address this and I believe either of these would serve my needs.

1. Add an option to resolve to a root-relative path. This would end up being something like the absolute option, but trimming the path to the root of the project or a specific subdirectory that represents the docroot for the website. [It looks like you attempted something with a root-relative URL](https://github.com/jkulhanek/gulp-resolve-url/blob/16a06d887b1b351d2ae4760eb421d00a9e1afa9f/index.js#L233) and I'm curious about your thinking there and why you abandoned it.
2. Add an option to resolve the paths relative to the destination. One would need to specify the same path passed to gulp.dest() in this option.

I think it would be simple to resolve urls relative to a valid path specified in either of those options [the way you currently calculate the relative path](https://github.com/jkulhanek/gulp-resolve-url/blob/16a06d887b1b351d2ae4760eb421d00a9e1afa9f/index.js#L232) with `path.relative(filePath, absolute)`, using the path specified in the option in place of `filePath`.

I'm willing to write the necessary changes, but before I proceed, do you have any thoughts on an existing way to accomplish this that I haven't considered? And would you consider incorporating a change that does either or both of these things, if I'm successful? Do you have any preliminary input on implementation?

**Comment by jkulhanek** [comment]
Ok, so you want to source files from a directory and make their paths relative to another directory? Is that right? Why don't you use the "root" option? Would it solve your problem?

In any case, could you please provide a minimal case (source code of what you want to achieve). It would be best if it was in a form of a unit-test. After that I can see, what I can do. In case of any modifications to this repo, I require test-driven-development and each change has to have a failing test associated with it, that that change fixes.

**Comment by tmanhollan** [comment]
I'm sorry for not responding sooner, I got caught up in another project right after I opened this issue.

The root option doesn't have any impact on my case. If I'm understating it correctly, the root option provides an additional place for resolveUrl to look for the file referenced in the url -- it's a hint for finding the file, not for how its path should be transformed. In my case, there is no issue with finding the files, which is evident by the fact that with the absolute option enabled, the resolved paths include the correct full filesystem path of the file referenced by the source.

If the purpose of the gulp task is to compile the sass files for use on a website, the issue is that after resolveUrl finishes, the stream will eventually be piped to a gulp.dest which resolveUrl is unaware of. It is the urls that end up in that output file that need to be properly resolved to be useful on the website. In a simple project layout where the output is in a parallel location to the source, there is no difference in their relative locations, so the resolved urls are correct when they eventually get written to the output.

I'll work on some code to demonstrate, but in the meantime, let me describe the issue using your original unit test. Consider the existing layout used in the unit tests, which is as follows (relative to the "test" directory):

fixtures
 - sub
   - module.scss
   - resource.png
 - main.scss
 - main.css

resolveUrl properly transforms the reference to "resource.png" in module.scss into "sub/resource.png" relative to the source file that's being compiled, main.scss. The url "sub/resource.png" is valid relative to the source (main.scss) and the output (main.css).

Consider the following modification to your directory structure:

fixtures
 - build
   - css
     - main.css
 - sub
   - module.scss
   - resource.png
 - main.scss

With gulp.dest("fixtures/build/css") and all other things being equal, the relative path "sub/resource.png" is valid from the persective of the source. But from the perspective of the output (main.css), the correct path should be "../../sub/resource.png".

The additional directory structure may seem overly complex in the context of the simple test structure, but it demonstrates the problem. In my case, the project structure is somewhat more complex, but the motivation for the output being inside the "build" directory is to keep the build artifacts out of source control and to keep the development process separate from the deployment process.

There are a number of other ways I could solve this by performing additional manipulations before or after the resolveUrl step, but your module's approach seems the most intelligent by trying to determine the intent of the original author; it's just missing one key element. My proposal is to simply make the module aware of the ultimate destination of the files so it can correct for it. The two approaches I originally described both achieve the same thing, but from different points of reference.

I'll work out some code and a test case when I get back to the project where I was using it. I'll submit what I come up with for review and then we can discuss it further if changes are necessary. Thanks for considering.

**Comment by jkulhanek** [comment]
Now I understand your problem completely. I think it could be achieved by a slight modification of the line 232 of file https://github.com/jkulhanek/gulp-resolve-url/blob/master/index.js, although I have not tested it yet. As soon as I have more time, I will look at it more carefully.

**Comment by tmanhollan** [comment]
Yes, that's the area I was thinking could be modified as well. I think there needs to be an additional option to specify the destination and some validation of it similar to the root path validation on line 56.

**Comment by tmanhollan** [comment]
I've added some code that addresses both approaches in my original description. I'm going to open a pull request for you to review, but consider it a work-in-progress. In particular I think the second approach (8b7e3d0) needs some work. I wanted to try both approaches because, although the second may be a bit more convoluted, I think it provides more flexibility. At the current state of my project, either approach works, but I can see down the road where the second may be more useful and where it may be useful for others as well. I'll devote some more attention to it next week, but feel free to add your suggestions when you have time.
